### PR TITLE
Graceful shutdown of zoo when daemon process terminates abruptly

### LINKED ISSUE #895: Processes not terminated
## **BUG REPORT**
**High Level Description**
I use SMARTS for training imitation learning agents with NGSIM dataset. When training is finished, SMARTS will leave some unterminated processes.

**SMARTS version**
0.4.15

**Steps to reproduce the bug**
I use the `multiprocessing` package to open multiple SMARTS instances for parallel sampling.

And I created each simulator using:

```python
self.smarts = SMARTS(
    agent_interfaces={},
    traffic_sim=None,
    visdom=None,
    envision=None,
    zoo_addrs=None,
)
```

**Resulting and expected behaviour**
There are multiple processes like `smarts/zoo/worker.py --port xxxxx` not terminated after the training stops.

**Error logs and screenshots**
截屏2021-06-02 11 16 23
 
**System information**
 
- Operating system & Environment [Please note if using Docker]: Ubuntu 16.04
- Python version [Please note if using Jupyter] 3.7.3
- SUMO version 1.8.0

### PR DESCRIPTION
**Pull request by Adaickalavan**
Caught abrupt terminate signals, in order to shutdown zoo manager and zoo workers gracefully.

Closes #895 .

### CODE REVIEW

**Review by RutvikGupta — APPROVED** [review]

**Comment by sah-huawei** [comment]
Although I think this is a good solution, when I tested it just now (by running `make test` and then hitting control-C about 40% of the way through), I got the following:
```
...
[gw0] [ 50%] PASSED smarts/core/tests/test_bubble_hijacking.py::test_bubble_hijacking 
smarts/core/tests/test_done_criteria.py::test_agents_alive_done_check[sim0] ^CTraceback (most recent call last):
  File "/home/steve/Repos/Huawei/SMARTS/smarts/zoo/manager.py", line 74, in <module>
    serve(args.port)
  File "/home/steve/Repos/Huawei/SMARTS/smarts/zoo/manager.py", line 58, in serve
    server.wait_for_termination()
  File "/home/steve/Repos/Huawei/SMARTS/.venv/lib/python3.7/site-packages/grpc/_server.py", line 976, in wait_for_termination
    timeout=timeout)
  File "/home/steve/Repos/Huawei/SMARTS/.venv/lib/python3.7/site-packages/grpc/_common.py", line 138, in wait
    _wait_once(wait_fn, MAXIMUM_WAIT_TIMEOUT, spin_cb)
  File "/home/steve/Repos/Huawei/SMARTS/.venv/lib/python3.7/site-packages/grpc/_common.py", line 103, in _wait_once
    wait_fn(timeout=timeout)
  File "/usr/lib/python3.7/threading.py", line 552, in wait
    signaled = self._cond.wait(timeout)
  File "/usr/lib/python3.7/threading.py", line 300, in wait
    gotit = waiter.acquire(True, timeout)
  File "/home/steve/Repos/Huawei/SMARTS/smarts/zoo/manager.py", line 47, in stop_server
    manager_servicer_object.destroy()
  File "/home/steve/Repos/Huawei/SMARTS/smarts/zoo/manager_servicer.py", line 94, in destroy
    for worker in self._workers.values():
RuntimeError: dictionary changed size during iteration
^C
```

Andt then there was still 1 zoo worker left running.

I think the reason is because `destroy()` (indirectly) calls `stop_worker()`, which modifies the `self._workers` dictionary.  If so, I think you can fix that by making a copy of the dict keys in `destroy()` to iterate over.

**Comment by Adaickalavan** [comment]
@sah-huawei: This pull request was meant to solve issue #895 .

**Comment by sah-huawei** [comment]
> @sah-huawei: This pull request was meant to solve issue #895 .

Yeah, and I think this exception in `destroy()` is probably (part of) why zoo processes are not terminated cleanly.  As I wrote, I still had a zombie zoo process after this happened.

File: smarts/core/remote_agent_buffer.py [hunk 1]
```diff
@@ -93,6 +93,16 @@ def __init__(self, zoo_manager_addrs=None, buffer_size=3, max_workers=4):
...
 
+        # Catch abrupt terminate signals
+        signal.signal(signal.SIGTERM, self._stop_servers)
+
+    def _stop_servers(self, *args):
+        self.destroy()
+        self._log.debug(
+            f"Shutting down zoo manager and zoo workers gracefully due to abrupt process stop."
+        )
+        sys.exit(0)
```
Comment (Gamenot) [inline smarts/core/remote_agent_buffer.py:104]: I had a small chuckle seeing gracefully on one line and `sys.exit()` immediately following.

File: smarts/core/remote_agent_buffer.py [hunk 2]
```diff
@@ -93,6 +93,16 @@ def __init__(self, zoo_manager_addrs=None, buffer_size=3, max_workers=4):
             self._remote_agent_future() for _ in range(self._buffer_size)
         ]
 
+        # Catch abrupt terminate signals
+        signal.signal(signal.SIGTERM, self._stop_servers)
```
Comment (Gamenot) [inline smarts/core/remote_agent_buffer.py:97]: Is there any need to unset the signal when the termination handler is called?

**Review by sah-huawei — APPROVED** [review]
Zhengbang pinged me on this one, so I went ahead and made the change I felt was needed.  It appears to have fixed the problem (for me at least) as now there are no zombies left after killing `make test`.

**Comment by Gamenot** [comment]
Please merge this in today.

File: smarts/core/remote_agent_buffer.py (hunk 1 above)
Comment (Adaickalavan) [reply smarts/core/remote_agent_buffer.py:106]: Removed the phrase `gracefully`.

File: smarts/core/remote_agent_buffer.py (hunk 2 above)
Comment (Adaickalavan) [reply smarts/core/remote_agent_buffer.py:99]: Probably. Need investigation on multiple calls to terminate.

**Pull request merged**